File: backend/app/routers/memory.py
import logging

from fastapi import APIRouter, HTTPException
from pydantic import BaseModel

from app import state

logger = logging.getLogger(__name__)

router = APIRouter(prefix="/api/v1/memory", tags=["memory"])


class RecallRequest(BaseModel):
    query: str
    device_id: str
    top_k: int = 5


class SearchRequest(BaseModel):
    query: str
    device_id: str
    top_k: int = 10


class FeedbackRequest(BaseModel):
    episode_id: str
    device_id: str
    rating: float


class ProfileInteractionRequest(BaseModel):
    interaction: dict


class EpisodeRequest(BaseModel):
    device_id: str
    query: str
    answer: str
    agents: list = []
    model_used: str = ""
    session_type: str = "chat"


class FactRequest(BaseModel):
    device_id: str
    content: str
    source_type: str = "conversation"
    source_id: str = ""


def _require_memory():
    if state.memory_service is None:
        raise HTTPException(status_code=503, detail="Memory service not available")
    return state.memory_service


@router.post("/recall")
async def recall(request: RecallRequest):
    svc = _require_memory()
    episodes = await svc.recall_episodes(request.device_id, request.query, top_k=request.top_k)
    facts = await svc.recall_facts(request.device_id, request.query)
    profile = await svc.get_profile(request.device_id)
    return {"episodes": episodes, "facts": facts, "profile": profile}


@router.get("/profile/{device_id}")
async def get_profile(device_id: str):
    svc = _require_memory()
    profile = await svc.get_profile(device_id)
    return {"device_id": device_id, "profile": profile}


@router.post("/profile/{device_id}")
async def update_profile(device_id: str, request: ProfileInteractionRequest):
    svc = _require_memory()
    profile = await svc.update_profile(device_id, request.interaction)
    return {"profile": profile}


@router.post("/episode")
async def create_episode(request: EpisodeRequest):
    svc = _require_memory()
    episode_id = await svc.store_episode(
        device_id=request.device_id,
        query=request.query,
        answer=request.answer,
        agents=request.agents,
        model_used=request.model_used,
        session_type=request.session_type,
    )
    return {"episode_id": episode_id}


@router.get("/episodes/{device_id}")
async def list_episodes(device_id: str, limit: int = 20, offset: int = 0):
    svc = _require_memory()
    result = await svc.list_episodes(device_id, limit=limit, offset=offset)
    return {"episodes": result, "total": len(result)}


@router.delete("/episode/{episode_id}")
async def delete_episode(episode_id: str, device_id: str):
    svc = _require_memory()
    episode = await svc.get_episode(episode_id)
    if not episode:
        raise HTTPException(status_code=404, detail="Episode not found")
    if episode.get("device_id") != device_id:
        raise HTTPException(status_code=403, detail="Forbidden: You do not own this episode")
    deleted = await svc.delete_episode(episode_id)
    return {"deleted": deleted}


@router.post("/fact")
async def create_fact(request: FactRequest):
    svc = _require_memory()
    fact_id = await svc.store_fact(
        device_id=request.device_id,
        content=request.content,
        source_type=request.source_type,
        source_id=request.source_id,
    )
    return {"fact_id": fact_id}


@router.get("/facts/{device_id}")
async def list_facts(device_id: str, query: str = "", top_k: int = 10):
    svc = _require_memory()
    result = await svc.recall_facts(device_id, query=query, top_k=top_k)
    return {"facts": result}


@router.get("/stats/{device_id}")
async def get_stats(device_id: str):
    svc = _require_memory()
    stats = await svc.get_stats(device_id)
    return stats


@router.get("/usage/{device_id}")
async def get_usage(device_id: str, metric: str | None = None, hours: int = 24):
    svc = _require_memory()
    usage = await svc.get_usage(device_id, metric_name=metric, hours=hours)
    return {"device_id": device_id, "usage": usage}


@router.post("/decay")
async def run_decay():
    svc = _require_memory()
    decayed = await svc.decay_old_facts()
    compacted = await svc.compact_episodes()
    return {"decayed_facts": decayed, "compacted_episodes": compacted}


@router.post("/search")
async def search(request: SearchRequest):
    svc = _require_memory()
    episodes = await svc.recall_episodes(request.device_id, request.query, top_k=request.top_k)
    facts = await svc.recall_facts(request.device_id, request.query)
    return {"episodes": episodes, "facts": facts}


@router.get("/history")
async def get_history(device_id: str, limit: int = 20, offset: int = 0):
    svc = _require_memory()
    result = await svc.list_episodes(device_id, limit=limit, offset=offset)
    return {"episodes": result, "total": len(result)}


@router.post("/feedback")
async def submit_feedback(request: FeedbackRequest):
    svc = _require_memory()
    if not 0.0 <= request.rating <= 5.0:
        raise HTTPException(status_code=400, detail="Rating must be between 0.0 and 5.0")

    episode = await svc.get_episode(request.episode_id)
    if not episode:
        raise HTTPException(status_code=404, detail="Episode not found")
    if episode.get("device_id") != request.device_id:
        raise HTTPException(status_code=403, detail="Forbidden: You do not own this episode")

    db = await svc._get_db()
    await db.execute(
        "UPDATE episodes SET outcome_rating = ? WHERE id = ?",
        (request.rating, request.episode_id),
    )
    await db.commit()
    return {"updated": True, "episode_id": request.episode_id, "rating": request.rating}


@router.get("/health")
async def memory_health():
    svc = state.memory_service
    if svc is None:
        return {"status": "unavailable", "db_exists": False}
    try:
        stats = await svc.get_stats()
        return {"status": "healthy", "db_exists": True, **stats}
    except Exception as e:
        return {"status": "error", "error": str(e), "db_exists": False}

File: frontend/lib/memory.ts
import { secureFetch, API_BASE_URL } from "./config";

const MEMORY_API = `${API_BASE_URL}/memory`;

export interface MemoryEpisode {
  id: string;
  query: string;
  answer: string;
  session_type: string;
  model_used: string;
  agents: string[];
  score?: number;
  created_at: number;
}

export interface MemoryFact {
  id: string;
  content: string;
  confidence: number;
  source_type: string;
  access_count: number;
}

export interface UserProfile {
  device_id: string;
  profile: Record<string, unknown>;
}

export interface MemoryRecall {
  episodes: MemoryEpisode[];
  facts: MemoryFact[];
  profile: UserProfile | null;
}

export interface MemoryStats {
  episodes: number;
  facts: number;
  profiles: number;
  total_dead_ends: number;
  total_strategies: number;
  usage_7d: Record<string, number>;
}

export async function recallMemory(query: string, deviceId: string, topK: number = 5): Promise<MemoryRecall> {
  const res = await secureFetch(`${MEMORY_API}/recall`, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({ query, device_id: deviceId, top_k: topK }),
  });
  if (!res.ok) throw new Error(`Memory recall failed: ${res.status}`);
  return res.json();
}

export async function getProfile(deviceId: string): Promise<UserProfile> {
  const res = await secureFetch(`${MEMORY_API}/profile/${deviceId}`);
  if (!res.ok) throw new Error(`Profile fetch failed: ${res.status}`);
  return res.json();
}

export async function updateProfile(deviceId: string, interaction: Record<string, unknown>): Promise<UserProfile> {
  const res = await secureFetch(`${MEMORY_API}/profile/${deviceId}`, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({ interaction }),
  });
  if (!res.ok) throw new Error(`Profile update failed: ${res.status}`);
  return res.json();
}

export async function storeEpisode(episode: {
  device_id: string;
  query: string;
  answer: string;
  agents?: string[];
  model_used?: string;
  session_type?: string;
}): Promise<{ episode_id: string }> {
  const res = await secureFetch(`${MEMORY_API}/episode`, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify(episode),
  });
  if (!res.ok) throw new Error(`Episode store failed: ${res.status}`);
  return res.json();
}

export async function getEpisodes(deviceId: string, limit: number = 20, offset: number = 0) {
  const res = await secureFetch(`${MEMORY_API}/episodes/${deviceId}?limit=${limit}&offset=${offset}`);
  if (!res.ok) throw new Error(`Episodes fetch failed: ${res.status}`);
  return res.json();
}

export async function deleteEpisode(episodeId: string, deviceId: string): Promise<{ deleted: boolean }> {
  const res = await secureFetch(`${MEMORY_API}/episode/${episodeId}?device_id=${deviceId}`, { method: "DELETE" });
  if (!res.ok) throw new Error(`Episode delete failed: ${res.status}`);
  return res.json();
}

export async function storeFact(fact: {
  device_id: string;
  content: string;
  source_type?: string;
  source_id?: string;
}): Promise<{ fact_id: string }> {
  const res = await secureFetch(`${MEMORY_API}/fact`, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify(fact),
  });
  if (!res.ok) throw new Error(`Fact store failed: ${res.status}`);
  return res.json();
}

export async function searchFacts(deviceId: string, query: string = "", topK: number = 10) {
  const params = new URLSearchParams();
  if (query) params.set("query", query);
  params.set("top_k", String(topK));
  const res = await secureFetch(`${MEMORY_API}/facts/${deviceId}?${params}`);
  if (!res.ok) throw new Error(`Facts search failed: ${res.status}`);
  return res.json();
}

export async function searchMemory(query: string, deviceId: string, topK: number = 10): Promise<{episodes: MemoryEpisode[]; facts: MemoryFact[]}> {
  const res = await secureFetch(`${MEMORY_API}/search`, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({ query, device_id: deviceId, top_k: topK }),
  });
  if (!res.ok) throw new Error(`Memory search failed: ${res.status}`);
  return res.json();
}

export async function getMemoryStats(deviceId: string): Promise<MemoryStats> {
  const res = await secureFetch(`${MEMORY_API}/stats/${deviceId}`);
  if (!res.ok) throw new Error(`Stats fetch failed: ${res.status}`);
  return res.json();
}

export async function triggerDecay(): Promise<{ decayed_facts: number; compacted_episodes: number }> {
  const res = await secureFetch(`${MEMORY_API}/decay`, { method: "POST" });
  if (!res.ok) throw new Error(`Decay trigger failed: ${res.status}`);
  return res.json();
}
